File: leetcode/SortColors/sortColors.java
public class Solution {
    public void sortColors(int[] A) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        int len = A.length, i0 = -1, i1 = 0, i2 = len;
        while (i1 < i2) {
            if (A[i1] == 2) {
                A[i1] = A[--i2];
                A[i2] = 2;
            } else {
                if (A[i1] == 0) {
                    A[i1] = A[++i0];
                    A[i0] = 0;
                }
                i1++;
            }
        }
    }
}

File: leetcode/SortColors/sortColors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int i0 = -1, i2 = nums.size(), i = 0;
        while (i < i2) {
            if (nums[i] == 2) {
                nums[i] = nums[--i2];
                nums[i2] = 2;
            }
            if (nums[i] == 0) {
                nums[i++] = nums[++i0];
                nums[i0] = 0;
            }
            if (nums[i] == 1) i++;
        }
    }
};
